111.md

```diff
@@ -0,0 +1,242 @@
+      `cosmos:<hub_id_name>` ||
+      `bip122:<16 bytes genesis/fork hash>`;
+
+let caip10 = `${caip02}:<checksum_address>`;
+```
+
+### d) Info
+`info` is CAIP-10 and NIP-02/NIP-05 identifier string formatted as:
+ ```js
+ let info = `${caip10}:${username}`;
+ ```
+
+### e) Message
+Deterministic `message` to be signed by the wallet provider,
+```js
+let message = `Log into Nostr client as '${username}'\n\nIMPORTANT: Please verify the integrity and authenticity of connected Nostr client before signing this message\n\nSIGNED BY: ${caip10}`
+```
+
+### f) Signature
+[RFC-6979](https://datatracker.ietf.org/doc/html/rfc6979) compatible (ECDSA) deterministic `signature` calculated by the wallet provider using native keypair,
+```js
+let signature = wallet.signMessage(message);
+```
+
+### g) Salt
+`salt` is SHA-256 hash of the `info`, optional password and last **32 bytes** of signature string formatted as:
+  ```js
+  let salt = await sha256(`${info}:${password?password:""}:${signature.slice(68)}`);
+  ```
+  where, `signature.slice(68)` are the last 32 bytes of the deterministic ECDSA-derived Ethereum signature.
+
+### h) Key Derivation Function (KDF)
+HMAC-Based KDF `hkdf(sha256, inputKey, salt, info, dkLen = 42)` is used to derive the **42 bytes** long **hashkey** with inputs,
+
+- `inputKey` is SHA-256 hash of signature bytes,
+   ```js
+   let inputKey = await sha256(hexToBytes(signature.slice(2)));
+   ```
+
+- `info` is same as defined before, i.e.
+   ```js
+   let info = `${caip10}:${username}`;
+   ```
+
+- `salt` is same as defined before, i.e.
+   ```js
+   let salt = await sha256(`${info}:${password?password:""}:${signature.slice(68)}`);
+   ```
+
+- `dkLen` (Derived Key Length) is set to `42`,
+   ```js
+   let dkLen = 42;
+   ```
+   [FIPS 186-4 B.4.1](https://csrc.nist.gov/publications/detail/fips/186/4/final) requires hashkey length to be `>= n + 8`, where `n = 32` is the **bytelength** of the final `secp256k1` private key, such that `42 >= 32 + 8`.
+
+- `hashToPrivateKey()` function is FIPS 186-4 B.4.1 implementation to convert HKDF-derived hashkey to valid `secp256k1` keypair. This function is implemented in JavaScript library `@noble/secp256k1` as `hashToPrivateKey()`.
+
+   ```js
+   let hashKey = hkdf(sha256, inputKey, salt, info, dkLen = 42);
+   let privKey = secp256k1.utils.hashToPrivateKey(hashKey);
+   let pubKey = secp256k1.schnorr.getPublicKey(privKey);
+   ```
+
+## Architecture
+
+The resulting architecture of NIP-111 can be visually interpreted as follows:
+
+![](https://raw.githubusercontent.com/dostr-eth/resources/main/graphics/nip-xx.png)
+
+## Implementation Requirements
+
+- Connected Ethereum wallet Signer **MUST** be EIP-191 and RFC-6979 compatible.
+- The `message` **MUST** be string formatted as
+```
+`Log into Nostr client as '${username}'\n\nIMPORTANT: Please verify the integrity and authenticity of connected Nostr client before signing this message\n\nSIGNED BY: ${caip10}`
```

Comment:
I'm a little confused, is this "inspired by" Sign-In With Ethereum/CAIP-122 or is it meant to be conformant to one or both of those two specs? If the latter was intended, the message should probably conform the CAIP-122/EIP-4361 ABNF, which could include the entire `message` (with `\n`s removed) in the `statement` field, use the `salt` value for `nonce`, and `info` as first entry in the `resources` array (not really sure what to do with `inputKey`, as I'm unclear on its exact function on a cursory read-through).  If it's not too late to go in that direction, I think it might have some benefits, such as being displayed to metamask users in the familiar, locked-down "Sign-In With Ethereum" modal rather than as a generic "personal_sign" modal... which only displays when presented with a personal_sign `message` matching the ABNF :D 
![image](https://user-images.githubusercontent.com/37127325/230104506-48ccf087-70fc-4514-8f10-aab1e7bb04c3.png)

Reply:
Thanks @bumblefudge! We intended to conform to CAIP-122 as much as possible but we had to strip all the variable quantities from the `message` format to retain deterministic nature of key derivation. In some sense, this implementation is not really "Sign-In" but more of "Ephemeral KeyGen and then Sign-In", and hence the necessary deviation from the standard.

Reply:
Well, the problem with an ABNF-based syntax is that you can't deviate and still get interop with other SIWX libraries or take advantage of SIWX support built into Metamask! Can I suggest hard-coding into the spec and template rather than removing the values you don't need, so that the message can still conform to the ABNF and get displayed to the users as a SIWX message?  I would note that ephemeral keygen is already baked into the SIWX standard, and is being used for that exact usecase by most implementers (the generated ephemeral public key is usually included as a value in the `Resources` array, although this isn't really mentioned explicitly in the specification itself!)

Reply:
> I'm a little confused, is this "inspired by" Sign-In With Ethereum/CAIP-122 or is it meant to be conformant to one or both of those two specs?  If the latter was intended, the message should probably conform the CAIP-122/EIP-4361 ABNF, which could include the entire message (with `\n` removed) in the statement field, use the salt value for nonce, and info as first entry in the resources array (not really sure what to do with inputKey, as I'm unclear on its exact function on a cursory read-through).
 
@bumblefudge 🙏  
Should we remove CAIP122 ref as we can't pass full strict ABNF? or req all as new CAIP?

We actually started as basic signature request before upgrading to full CAIP122/ERC4361 format but had to remove all extradata for deterministic keys as they are too strict to fit in all.. So it's now back to generic "personal_sign" modal for chain agnostic "sign-in-with-x" in Nostr context, internally it's using deterministic signature from wallet to generate "app specific deterministic keys" across all Nostr clients.

We could fill in all ABNF *required* formats with deterministic/fixed values but we can't pass `URI` validation in wallets, and "`${service} wants you to sign in with <chain> account: \n<addr>\n...must:have`" text message for ALL web2+3 D/Apps is too strict for our deterministic keygen, web3+dapps are not supposed to have single URLs/apps as entry point.. 

"Important: .....\n\n" is supposed to be "Warning! ..." statement, it's not key: value. All other extradata is wrapped in `SIGNED BY : ${CAIP10}`, & our msg is simple 3 blocks `<title>\n\n<statement>\n\n<key:value>...` 


> I would note that ephemeral keygen is already baked into the SIWX standard, and is being used for that exact usecase by most implementers (the generated ephemeral public key is usually included as a value in the Resources array, although this isn't really mentioned explicitly in the specification itself!) 

Can you add some ref codes/links for that?

Reply:
> > I'm a little confused, is this "inspired by" Sign-In With Ethereum/CAIP-122 or is it meant to be conformant to one or both of those two specs?  If the latter was intended, the message should probably conform the CAIP-122/EIP-4361 ABNF, which could include the entire message (with `\n` removed) in the statement field, use the salt value for nonce, and info as first entry in the resources array (not really sure what to do with inputKey, as I'm unclear on its exact function on a cursory read-through).
> 
> @bumblefudge 🙏 Should we remove CAIP122 ref as we can't pass full strict ABNF? or req all as new CAIP?
> 
> We actually started as basic signature request before upgrading to full CAIP122/ERC4361 format but had to remove all extradata for deterministic keys as they are too strict to fit in all.. So it's now back to generic "personal_sign" modal for chain agnostic "sign-in-with-x" in Nostr context, internally it's using deterministic signature from wallet to generate "app specific deterministic keys" across all Nostr clients.
> 
> We could fill in all ABNF _required_ formats with deterministic/fixed values but we can't pass `URI` validation in wallets, and "`${service} wants you to sign in with <chain> account: \n<addr>\n...must:have`" text message for ALL web2+3 D/Apps is too strict for our deterministic keygen, web3+dapps are not supposed to have single URLs/apps as entry point..
> 
> "Important: .....\n\n" is supposed to be "Warning! ..." statement, it's not key: value. All other extradata is wrapped in `SIGNED BY : ${CAIP10}`, & our msg is simple 3 blocks `<title>\n\n<statement>\n\n<key:value>...`
> 
> > I would note that ephemeral keygen is already baked into the SIWX standard, and is being used for that exact usecase by most implementers (the generated ephemeral public key is usually included as a value in the Resources array, although this isn't really mentioned explicitly in the specification itself!)
> 
> Can you add some ref codes/links for that?

`URI` is definitely a bigger issue and it hadn't come to my mind before. While `timestamp` and `nonce` can be replaced with placeholders, `URI` is used by Metamask for its community-audited safe dApp list. This is problematic for dApps or services without a unique entry point. CAIP122 is too strict for all use-cases and pretty much a death sentence for the deterministic use-case.

Reply:
Couldn't the uri be:
https://github.com/nostr-protocol/nips/blob/master/111.md
? It doesn't need to be on the `domain`

Reply:
> Couldn't the uri be: https://github.com/nostr-protocol/nips/blob/master/111.md ? It doesn't need to be on the `domain`

That's not a bad idea honestly but it will make users highly suspicious and wary of a service with 1/1/1970 in timestamp and a GitHub link in URI. I personally won't sign and subscribe to such a service at first glance. I believe there is room for a new CAIP detailing a separate signature format standard for deterministic use-cases.

Reply:
> Couldn't the uri be: https://github.com/nostr-protocol/nips/blob/master/111.md ? It doesn't need to be on the `domain`

When we connect MM at `https://app.dostr.eth.limo` it'll check and add "WARNING" if connected domain isn't matching `domain` URL. It's good feature for web2 & web3 apps with single point of entry but for our use case we're adding extra user info & pw with static signature request as basic security.

It all work as plaintext "sign in with xyz on Nostr" signature request so we're missing all SWIx interface/features. 

+ it's better if we explain why we're doing this, our keygen design is inspired by Umbra cash & old Whisper(ssh) newKeypair/ newIdentity function. We use `window.ethereum` to sign eth tx/permits then use deterministic keys in `window.nostr` to send that over Nostr relays for off-chain features like stealth payments, DEFI, NST markets, alt-mempool for AA, xyz services, bundlers/bots, games...

Reply:
personally I think if we don't need timestamp, adding a placeholder for it is not a good idea.